server/neptune/workflows/internal/pr/runner.go

```diff
@@ -0,0 +1,196 @@
+// dealing with any failed policies by reviewing set of approvals
+func (r *Runner) processRevision(ctx workflow.Context, prRevision receiver.Revision) {
+	r.state = Working
+	defer func() {
+		r.lastProcessedRevision = prRevision.Revision
+		r.state = Waiting
+	}()
+	failedPolicies := sync.Map{}
+	var wg sync.WaitGroup
+	wg.Add(len(prRevision.Roots))
+	for _, root := range prRevision.Roots {
+		rootCtx := workflow.WithValue(ctx, internalContext.ProjectKey, root.Name)
+		workflow.Go(rootCtx, func(c workflow.Context) {
+			defer wg.Done()
+			failedRootPolicies, err := r.runTerraformWorkflow(c, root, prRevision)
+			if err != nil {
+				// choosing to not fail workflow and let it continue to exist
+				// until PR close/timeout
+				errCtx := workflow.WithValue(ctx, internalContext.ErrKey, err)
+				workflow.GetLogger(errCtx).Error("processing pr revision")
+			}
+			// consolidate failures across all roots
+			// policy sets are identical so multiple roots can fail the same policy without issue
+			for _, failedPolicy := range failedRootPolicies {
+				failedPolicies.LoadOrStore(failedPolicy.Name, failedPolicy)
+			}
+		})
+	}
+	wg.Wait()
+	// TODO: check for policy failures
+}
+
+func (r *Runner) runTerraformWorkflow(ctx workflow.Context, root terraform.Root, prRevision receiver.Revision) ([]activities.PolicySet, error) {
+	id, err := sideeffect.GenerateUUID(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "generating uuid")
+	}
+	prRootInfo := internalTerraform.PRRootInfo{
+		ID: id,
+		Commit: github.Commit{
+			Revision: prRevision.Revision,
+		},
+		Root: root,
+		Repo: prRevision.Repo,
+	}
+	failedPolicies, err := r.TFWorkflowRunner.Run(ctx, prRootInfo)
+	if err != nil {
+		return nil, errors.Wrap(err, "running terraform workflow")
+	}
+	return failedPolicies, err
+}
+
+func (r *Runner) shouldProcessRevision(ctx workflow.Context, prRevision receiver.Revision) (bool, error) {
+	if r.state == Working {
+		return false, nil
+	}
+	if r.lastProcessedRevision == "" {
+		return true, nil
+	}
+	direction, err := r.getCommitDirection(ctx, prRevision)
+	if err != nil {
+		return false, err
+	}
+	// TODO: should we handle diverged?
+	return direction != activities.DirectionBehind, nil
+}
+
+func (r *Runner) getCommitDirection(ctx workflow.Context, prRevision receiver.Revision) (activities.DiffDirection, error) {
+	// this means we are deploying this root for the first time
+	if r.lastProcessedRevision == "" {
+		return activities.DirectionAhead, nil
+	}
+	var compareCommitResp activities.CompareCommitResponse
+	err := workflow.ExecuteActivity(ctx, r.Activities.GithubCompareCommit, activities.CompareCommitRequest{
+		DeployRequestRevision:  prRevision.Revision,
+		LatestDeployedRevision: r.lastProcessedRevision,
+		Repo:                   prRevision.Repo,
+	}).Get(ctx, &compareCommitResp)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to determine new revision commit direction")
+	}
+	return compareCommitResp.CommitComparison, nil
+}
```

Comment:
I'm not convinced that we need this, if we just fetch the latest commit for the PR when we get signaled.  In the deploy case, we don't necessarily want to do that because we want to deploy revisions 1 at a time. In the pr case, that doesn't matter.